README.md

```diff
@@ -37,6 +40,58 @@ export default (workspaceDir) => {
 }
 ```
 
+### Example 2
+
+The following `.meta-updater/main.mjs` defines format `#ignore` and updaters for `.gitignore` and `.eslintrc` using explicit format specifier.
+
+```js
+import * as fs from 'fs/promises'
+import { createFormat, createUpdateOptions } from '@pnpm/meta-updater'
```

Comment:
Why do you need these functions?

You could just export a new field from the update file. Like

```
export const formats = {
   ...
}
```

Reply:
Both functions are used to constraint types.

## Ways to use meta-updater

#### Unsafe: JavaScript without IDE/without typescript language server

No need to use `createFormat` / `createUpdateOptions` at all. You can see it in action in `__fixtures__/workspace-1/.meta-updater/main.mjs` - it is plain object without `createUpdateOptions`.

#### Unsafe: JavaScript with IDE/with typescript language server

Without manually annotating with JSDocs and without `createFormat` / `createUpdateOptions` all the parameters in file updaters will be inferred as `any`. Nothing stops from using meta-updater like this.

#### Safe: TypeScript or JavaScript with IDE/with typescript language server

Everything should be annotated with types (TS) or JSDocs (JS) or `createFormat` / `createUpdateOptions` should be used.

## Format approaches - Safe only

#### Without `createFormat` in TypeScript

pros:
- no use of createFormat / createUpdateOptions

cons:
- need to annotate everything with types every time for every user (not DRY)
- mental load - need to know where content is of type `string[]` and where it is of type `string[] | null`

```ts
const gitignoreFormat = {
  async read({ resolvedPath }: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) {
    return (await fs.readFile(resolvedPath, 'utf8')).split('\n')
  },
  update(actual: string[] | null, updater: (content: string[] | null, options: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) => null | string[] | Promise<null | string[]>, options: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) {
    return updater(actual, options)
  },
  equal(expected: string[], actual: string[]) {
    return equals(expected, actual)
  },
  async write(expected: string[], { resolvedPath }: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) {
    const unique = <T extends unknown[]>(array: T) => Array.from(new Set<T[number]>(array)).sort()
    await fs.writeFile(resolvedPath, unique(expected).join('\n'), 'utf8')
  },
}
```

#### Without `createFormat` in JavaScript

JavaScript

```js
const gitignoreFormat = {
  /**
   * @param {{dir: string, manifest: import('@pnpm/types').ProjectManifest}} param0 
   */
  async read({ resolvedPath }) {
    return (await fs.readFile(resolvedPath, 'utf8')).split('\n')
  },
  /**
   * @param {string[] | null} actual
   * @param {{(content: string[] | null, options: {dir: string, manifest: import('@pnpm/types').ProjectManifest}): null | string[] | Promise<null | string[]>}} updater
   * @param {{dir: string, manifest: import('@pnpm/types').ProjectManifest}} param2 
   */
   update(actual, updater, options) {
    return updater(actual, options)
  },
  /**
   * 
   * @param {string[]} expected 
   * @param {string[]} actual 
   * @returns 
   */
  equal(expected, actual) {
    return R.equals(expected, actual)
  },
  /**
   * @param {string[]} expected
   * @param {{dir: string, manifest: import('@pnpm/types').ProjectManifest}} param1 
   */
   async write(expected, { resolvedPath }) {
    const unique = (array) => Array.from(new Set() < T[number] > array).sort()
    await fs.writeFile(resolvedPath, unique(expected).join('\n'), 'utf8')
  },
}
```

#### Format definition with `createFormat`:

pros:
- works in both TypeScript and JavaScript with zero type annotations
- type of format content is inferred from return type of the `read` method and contextually typed in all other locations:
- types of parameters are contextually typed

cons:
- runtime overhead for basically identity function (createFormat compiled to JS looks like: `const createFormat = a => a`)

```js
const gitignoreFormat = createFormat({
  async read({ resolvedPath }) {
    return (await fs.readFile(resolvedPath, 'utf8')).split('\n')
  },
  update(actual, updater, options) {
    return updater(actual, options)
  },
  equal(expected, actual) {
    return equals(expected, actual)
  },
  async write(expected, { resolvedPath }) {
    const unique = (array) => Array.from(new Set(array)).sort()
    await fs.writeFile(resolvedPath, unique(expected).join('\n'), 'utf8')
  },
})
```

## How type safety looks (screenshots are from JavaScript main.mjs):

#### files
![image](https://user-images.githubusercontent.com/1188919/190078306-c908fa1f-9f85-4957-9314-ed5c7959c8b8.png)

![image](https://user-images.githubusercontent.com/1188919/190078388-11aa4562-8b57-445c-baf4-26db5dc8f0a1.png)

![image](https://user-images.githubusercontent.com/1188919/190078142-453d0e74-0d0c-43d5-a2af-0d31c024141b.png)

![image](https://user-images.githubusercontent.com/1188919/190078217-6bf0ca69-2911-485e-9b61-145a219e96e7.png)

![image](https://user-images.githubusercontent.com/1188919/190078492-56ef0403-8303-4ad6-9a9b-bf6cafa8a9ca.png)

#### format

![image](https://user-images.githubusercontent.com/1188919/190081290-4e1f54ba-a259-4a03-b4c5-09371e3cd346.png)

![image](https://user-images.githubusercontent.com/1188919/190081387-0e1745e2-6316-4be6-a4ff-09bc5d9ff7f9.png)